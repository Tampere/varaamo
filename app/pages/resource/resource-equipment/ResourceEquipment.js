import React from 'react';
import PropTypes from 'prop-types';
import Col from 'react-bootstrap/lib/Col';
import Row from 'react-bootstrap/lib/Row';

import injectT from '../../../i18n/injectT';
import ResourcePanel from '../resource-info/ResourcePanel';

function ResourceEquipment({
  equipment = [],
  t,
}) {
  const equipmentColumns = equipment.map(
    (item, i) => <Col key={i} lg={3} md={3} xs={6}>{item.name}</Col>,
  );
  return (
    <ResourcePanel header={t('ResourceEquipment.headingText')}>
      <Row>
        {equipmentColumns}
      </Row>
    </ResourcePanel>
  );
}

ResourceEquipment.propTypes = {
  equipment: PropTypes.array.isRequired,
  t: PropTypes.func.isRequired,
};

export default injectT(ResourceEquipment);
