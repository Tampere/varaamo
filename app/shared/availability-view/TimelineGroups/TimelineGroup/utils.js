import { slotSize, slotWidth, slotMargin } from 'constants/SlotConstants';

import some from 'lodash/some';
import moment from 'moment';

function getTimeSlotWidth({ startTime, endTime } = {}) {
  const diff = endTime ? endTime.diff(startTime, 'minutes') : slotSize;
  const slots = Math.floor(diff / slotSize);

  return (slotWidth * slots) - slotMargin;
}

function getTimelineItems(date, reservations, resourceId) {
  const items = [];
  let reservationPointer = 0;
  let timePointer = date.clone().startOf('day');
  const end = date.clone().endOf('day');
  while (timePointer.isBefore(end)) {
    const reservation = reservations && reservations[reservationPointer];
    const isSlotReservation = reservation && timePointer.isSame(reservation.begin);
    if (isSlotReservation) {
      items.push({
        key: String(items.length),
        type: 'reservation',
        data: reservation,
      });
      timePointer = moment(reservation.end);
      reservationPointer += 1;
    } else {
      items.push({
        key: String(items.length),
        type: 'reservation-slot',
        data: {
          begin: timePointer.format(),
          end: timePointer.clone().add(slotSize, 'minutes').format(),
          resourceId,
          // isSelectable: false by default to improve selector performance by allowing
          // addSelectionData to make some assumptions.
          isSelectable: false,
        },
      });
      timePointer.add(slotSize, 'minutes');
    }
  }
  return items;
}

function isInsideOpeningHours(item, openingHours) {
  return some(openingHours, opening => (
    opening.opens <= item.data.begin && item.data.end <= opening.closes
  ));
}

function markItemSelectable(item, isSelectable, openingHours) {
  const selectable = (
    isSelectable
    && moment().isSameOrBefore(item.data.end)
    && (!openingHours || isInsideOpeningHours(item, openingHours))
  );
  return { ...item, data: { ...item.data, isSelectable: selectable } };
}

function markItemsSelectable(items, isSelectable, openingHours) {
  return items.map((item) => {
    if (item.type === 'reservation') return item;
    return markItemSelectable(item, isSelectable, openingHours);
  });
}

function addSelectionData(selection, resource, items) {
  if (!selection) {
    return markItemsSelectable(items, true, resource.openingHours);
  } if (selection.resourceId !== resource.id) {
    // isSelectable is false by default, so nothing needs to be done.
    // This is a pretty important performance optimization when there are tons of
    // resources in the AvailabilityView and the selection is in a state where the
    // first click has been done but the second (end time) hasn't. Without this
    // optimization we'd be calling markItemSelectable for every slot in every
    // resource when the user hovers to another slot.
    return items;
  }
  let lastSelectableFound = false;
  return items.map((item) => {
    if (lastSelectableFound || item.data.begin < selection.begin) {
      if (item.type === 'reservation') return item;
      // isSelectable is false by default.
      return item;
    }
    if (item.type === 'reservation') {
      lastSelectableFound = true;
      return item;
    }
    return markItemSelectable(item, true, resource.openingHours);
  });
}

export default {
  addSelectionData,
  getTimelineItems,
  getTimeSlotWidth,
};
