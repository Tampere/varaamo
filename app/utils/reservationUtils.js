import camelCase from 'lodash/camelCase';
import clone from 'lodash/clone';
import find from 'lodash/find';
import last from 'lodash/last';
import some from 'lodash/some';
import sortBy from 'lodash/sortBy';
import tail from 'lodash/tail';
import get from 'lodash/get';
import moment from 'moment';

import constants from '../constants/AppConstants';

function combine(reservations) {
  if (!reservations || !reservations.length) {
    return [];
  }

  const sorted = sortBy(reservations, 'begin');
  const initialValue = [clone(sorted[0])];

  return tail(sorted).reduce((previous, current) => {
    if (current.begin === last(previous).end) {
      last(previous).end = current.end;
    } else {
      previous.push(clone(current));
    }
    return previous;
  }, initialValue);
}

function isStaffEvent(reservation, resource) {
  if (!resource || !resource.requiredReservationExtraFields) {
    return false;
  }
  return some(resource.requiredReservationExtraFields, field => (
    !reservation[camelCase(field)]
  ));
}

function getCurrentReservation(reservations) {
  const now = moment();
  return find(
    reservations, reservation => moment(reservation.begin) < now && now < moment(reservation.end),
  );
}

function getMissingValues(reservation) {
  const missingValues = {};
  constants.REQUIRED_STAFF_EVENT_FIELDS.forEach((field) => {
    if (!reservation[field]) {
      missingValues[field] = '-';
    }
  });
  return missingValues;
}

function getNextAvailableTime(reservations, fromMoment = moment()) {
  const combinedReservations = combine(reservations);
  if (!combinedReservations.length || fromMoment < moment(combinedReservations[0].begin)) {
    return fromMoment;
  }
  const ongoingReservation = find(combinedReservations, reservation => (
    moment(reservation.begin) <= fromMoment && fromMoment < moment(reservation.end)
  ));
  return ongoingReservation ? moment(ongoingReservation.end) : fromMoment;
}

function getNextReservation(reservations) {
  const now = moment();
  const orderedReservations = sortBy(reservations, reservation => moment(reservation.begin));
  return find(orderedReservations, reservation => now < moment(reservation.begin));
}

// A reservation may be requested with the resource inlined. In these
// instances the id is contained in `resource.id` field.
function getReservationResourceId(resource) {
  if (!resource) {
    return undefined;
  }

  const resourceIsString = typeof resource === 'string';
  const isResourceId = resourceIsString;
  const isInlinedResource = !resourceIsString;

  if (isResourceId) {
    return resource;
  }

  if (isInlinedResource) {
    return resource.id;
  }

  return undefined;
}

function getEditReservationUrl(reservation) {
  const {
    begin, end, id, resource,
  } = reservation;
  const date = moment(begin).format('YYYY-MM-DD');
  const beginStr = moment(begin).format('HH:mm');
  const endStr = moment(end).format('HH:mm');
  const resourceId = getReservationResourceId(resource);

  return `/reservation?begin=${beginStr}&date=${date}&end=${endStr}&id=${id || ''}&resource=${resourceId}`;
}
/**
 * Get reservation price from resource. Get time conver
 *
 * @param {ApiClient} apiClient
 * @param {String} begin Begin timestamp in ISO string
 * @param {String} end End timestamp in ISO string
 * @param {Array} products Resource product data.
 * @returns {Promise<string|null} Price or no price.
 */
async function getReservationPrice(apiClient, begin, end, products) {
  const productId = get(products, '[0].id');
  if (!begin || !end || !productId) {
    return null;
  }
  try {
    const payload = {
      begin,
      end,
      order_lines: [{ product: productId }],
    };
    const result = await apiClient.post('order/check_price', payload);
    const price = get(result, 'data.price');
    return price;
  } catch (e) {
    return null;
  }
}

function getReservationPricePerPeriod(resource) {
  const price = get(resource, 'products[0].price.amount');
  const pricePeriod = get(resource, 'products[0].price.period');
  const priceType = get(resource, 'products[0].price.type');
  const duration = moment.duration(pricePeriod);
  const hours = duration.asHours();
  const period = hours >= 1
    ? `${hours} h`
    : `${duration.asMinutes()} min`;
  const priceEnding = priceType === 'fixed' ? '' : ` / ${period}`;

  return `${price}â‚¬${priceEnding}`;
}

export {
  combine,
  isStaffEvent,
  getCurrentReservation,
  getEditReservationUrl,
  getMissingValues,
  getNextAvailableTime,
  getNextReservation,
  getReservationResourceId,
  getReservationPrice,
  getReservationPricePerPeriod,
};
