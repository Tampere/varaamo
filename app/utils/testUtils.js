import { shallow } from 'enzyme';
import forEach from 'lodash/forEach';
import forIn from 'lodash/forIn';
import get from 'lodash/get';
import mapValues from 'lodash/mapValues';
import merge from 'lodash/merge';
import set from 'lodash/set';
import React from 'react';
import { IntlProvider } from 'react-intl';
import { RSAA } from 'redux-api-middleware';

import rootReducer from 'state/rootReducer';
import enMessages from 'i18n/messages/en.json';

const testMessages = mapValues(enMessages, (value, key) => key);

function createApiTest(options) {
  /** Create some tests for an API action.
   *
   * Options can be an object like this:
   * {
   *  name: name of test (string),
   *  action: action creator (function),
   *  args: arguments for the action creator (array),
   *  tests: {
   *    method: HTTP method (string),
   *    endpoint: URL (string),
   *    headers: request headers (object)
   *    body: request body (string)
   *    meta: meta of the action (object)
   *    request: {
   *      type: request action type (string),
   *      hasSideEffect: (boolean)
   *      extra: {
   *        tests: {
   *          (testName): ({ payload, promise, response }, (done)) (function)
   *        }
   *      }
   *    },
   *    success: {
   *      type: success action type (string),
   *      hasSideEffect: (boolean)
   *      payload: {
   *        data: payload response data (json-like)
   *        tests: {
   *          (testName): ({ payload, promise, response }, (done)) (function)
   *        }
   *      }
   *      extra: {
   *        tests: {
   *          (testName): ({ payload, promise, response }, (done)) (function)
   *        }
   *      }
   *    },
   *    error: {
   *      type: error action type (string),
   *      hasSideEffect: (boolean),
   *      extra: {
   *        tests: {
   *          (testName): ({ payload, promise, response }, (done)) (function)
   *        }
   *      }
   *    },
   *  },
   * }
   */
  const tests = options.tests;
  describe(options.name, () => {
    let action;
    let callAPI;

    beforeAll(() => {
      action = options.action.apply(null, options.args);
      callAPI = action[RSAA];
    });

    test('returns an object with a RSAA key', () => {
      expect(callAPI).toBeDefined();
    });

    tests.method && test(`uses ${tests.method}`, () => {
      expect(callAPI.method).toBe(tests.method);
    });

    tests.body && test('has correct body', () => {
      const body = (
        tests.body instanceof Object
          ? JSON.parse(callAPI.body)
          : callAPI.body
      );
      expect(body).toEqual(tests.body);
    });

    tests.endpoint && test('uses correct endpoint', () => {
      expect(callAPI.endpoint).toBe(tests.endpoint);
    });

    tests.meta && test('has correct meta', () => {
      expect(action.meta).toEqual(tests.meta);
    });

    describe('types', () => {
      const mockAction = {
        [RSAA]: {
          types: [{
            type: (
              (tests.request && tests.request.type) || 'Specify request.type'
            )
          }],
        },
      };

      ['request', 'success', 'error'].forEach((actionTypeName, index) => {
        tests[actionTypeName] && describe(actionTypeName, () => {
          const actionTests = tests[actionTypeName];
          const extraTests = actionTests.extra && actionTests.extra.tests;
          let typeAction;
          beforeAll(() => {
            typeAction = callAPI.types[index];
          });

          actionTests.type && test('has correct type', () => {
            expect(typeAction.type).toBe(actionTests.type);
          });

          extraTests && forIn(extraTests, (value, name) => {
            const getParams = () => ({
              typeAction,
              mockAction,
              meta: (
                actionTypeName === 'request'
                  ? typeAction.meta
                  : typeAction.meta(mockAction)
              ),
            });
            const func = (
              value.length === 2
                ? done => value(getParams(), done)
                : () => value(getParams())
            );
            test(name, func);
          });

          actionTypeName === 'success' && describe('payload', () => {
            let payload;
            let response;
            let promise;
            const payloadTests = actionTests.payload || {};
            const jsonData = payloadTests.data || {};

            beforeAll(() => {
              payload = typeAction.payload;
              response = {
                headers: {
                  get: () => 'application/json',
                },
                json: () => Promise.resolve(jsonData),
              };
            });

            beforeEach(() => {
              promise = payload({ type: actionTests.type }, {}, response);
            });

            test('exists', () => {
              expect(payload).toBeDefined();
            });

            payloadTests.tests && forIn(payloadTests.tests, (value, name) => {
              const func = (
                value.length === 2
                  ? done => value({ promise, payload, response }, done)
                  : () => value({ promise, payload, response })
              );
              test(name, func);
            });
          });
        });
      });
    });
  });
}

function getDefaultRouterProps() {
  return {
    location: {
      query: {},
    },
    params: {},
  };
}

function getInitialState() {
  return rootReducer(undefined, { type: 'NOOP' });
}

function getState(extraState = {}) {
  const state = getInitialState();
  const newState = {};

  forEach(Object.keys(extraState), (key) => {
    const mergedValue = merge({}, get(state, key), extraState[key]);
    set(newState, key, mergedValue);
  });
  return merge({}, state, newState);
}

function makeButtonTests(button, name, expectedText, expectedOnClickFunction) {
  test(`is an ${name} button`, () => {
    expect(button.props().children).toBe(expectedText);
  });

  test('clicking the button calls correct onClick function', () => {
    expectedOnClickFunction.reset();
    button.props().onClick();

    expect(expectedOnClickFunction.callCount).toBe(1);
  });
}

// Utilities for shallow testing components that use react-intl and injectT.
const intlProvider = new IntlProvider({ locale: 'en', messages: testMessages }, {});
const { intl } = intlProvider.getChildContext();

function shallowWithIntl(node, context) {
  const nodeWithIntl = React.cloneElement(node, { intl });
  return shallow(nodeWithIntl, { context: { ...context, intl } }).shallow({ context });
}

export {
  createApiTest,
  getDefaultRouterProps,
  getInitialState,
  getState,
  makeButtonTests,
  shallowWithIntl,
};
