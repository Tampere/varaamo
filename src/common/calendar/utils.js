import moment from 'moment';
import get from 'lodash/get';

/**
 * getDefaultSelectedTimeRange()
 * Get default selected time range from reservation data.
 * @returns {Object} selected time range
 */
export const getDefaultSelectedTimeRange = (reservation) => {
  return reservation ? {
    start: reservation.begin,
    end: reservation.end,
  } : null;
};


/**
 * isTimeRangeOverMaxPeriod();
 * Reservation cannot be longer than the resources max period if max period is set.
 *
 * @param {Object} resource Resource data
 * @param {Date} start Start time
 * @param {Date} end End time
 * @param {boolean} isStaff Bypass this check if user is staff member
 * @returns {boolean} Selected time range is larger than max_period or not
 */
export const isTimeRangeOverMaxPeriod = (resource, start, end, isStaff = false) => {
  const period = get(resource, 'max_period', null);

  if (!isStaff && period) {
    const startMoment = moment(start);
    const endMoment = moment(end);

    const periodDuration = moment.duration(period);
    const durationInMinutes = periodDuration.hours() * 60 + periodDuration.minutes();

    if (endMoment.diff(startMoment, 'minutes') > durationInMinutes) {
      return true;
    }
  }

  return false;
};

/**
 * isTimeRangeUnderMinPeriod();
 * Reservation must longer than the resources min period if min period is set.
 *
 * @param {Object} resource Resource data
 * @param {Date} start Start time
 * @param {Date} end End time
 * @param {boolean} isStaff Bypass this check if user is staff member
 * @returns {boolean} Selected time range is larger than min_period or not
 */
export const isTimeRangeUnderMinPeriod = (resource, start, end) => {
  const period = get(resource, 'min_period', null);

  if (period) {
    const startMoment = moment(start);
    const endMoment = moment(end);

    const periodDuration = moment.duration(period);
    const durationInMinutes = periodDuration.hours() * 60 + periodDuration.minutes();

    if (endMoment.diff(startMoment, 'minutes') < durationInMinutes) {
      return true;
    }
  }

  return false;
};

/**
 * addPeriodToTime()
 * Add period to input time
 * For example: period 00:30:00, time: moment(12:00:00) -> moment(12:30:00)
 *
 * @param {Moment} time
 * @param {string} period
 * @returns {Moment}
 */

const addPeriodToTime = (time, period) => {
  if (period) {
    const periodInMinute = moment.duration(period).as('minutes');
    return time.add(periodInMinute, 'minutes');
  }

  // Return original if no period
  return time;
};

/**
 * getMinPeriodTimeRange
 * Populate the end time with resource min_period
 * For example: start time 1pm, min_period: 1:00:00 ->
 * end time should be 1pm + 1hour = 2pm
 *
 * @param {Object} resource
 * @param {Date} start
 * @param {Date} end
 * @returns {Date} return new end date if time range < min_period given, otherwise return original end
 */

export const getMinPeriodTimeRange = (resource, start, end) => {
  const minPeriod = get(resource, 'min_period', null);

  if (!minPeriod) {
    return {
      start, end,
    };
  }

  const minPeriodEnd = addPeriodToTime(moment(start), minPeriod);

  if (minPeriodEnd.diff(moment(end)) > 0) {
    // time range smaller than min_period, return min_period
    return {
      start,
      end: minPeriodEnd.toDate(),
    };
  }

  return {
    start,
    end,
  };
};

/**
 * getMaxPeriodEndTime
 * Populate the end time with resource min_period
 * For example: start time 1pm, min_period: 1:00:00 ->
 * end time should be 1pm + 1hour = 2pm
 *
 * @param {Object} resource
 * @param {Date} start
 * @param {Date} end
 * @returns {Date} return new end date if time range < min_period given, otherwise return original end
 */

export const getMaxPeriodTimeRange = (resource, start, end, isStaff) => {
  const maxPeriod = get(resource, 'max_period', null);

  if (!maxPeriod || isStaff) {
    return {
      start, end,
    };
  }

  const maxPeriodEnd = addPeriodToTime(moment(start), maxPeriod);

  if (moment(end).diff(maxPeriodEnd) > 0) {
    // time range is over max_period not allowed, return max_period
    return {
      start,
      end: maxPeriodEnd.toDate(),
    };
  }

  return {
    start,
    end,
  };
};
