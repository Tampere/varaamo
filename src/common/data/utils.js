import get from 'lodash/get';
import values from 'lodash/values';
import pickBy from 'lodash/pickBy';
import camelCase from 'lodash/camelCase';

export const getLocalizedFieldValue = (field, locale, fallback = false) => {
  const localeValue = get(field, locale, null);

  if (localeValue || !fallback) {
    return localeValue;
  }

  return values(field).find(fallbackValue => !!fallbackValue);
};

/**
 * Trim empty values from data object
 *
 * @param {Object} data
 * @returns {Object} Trimmed empty values object
 */

export const parseData = (data) => {
  return pickBy(data, value => value || value === 0);
};

export const toCamelCase = (obj) => {
  if (!obj) return {};

  const camelCasedObj = {};

  Object.keys(obj).forEach((key) => {
    const camelCasedKey = camelCase(key);
    camelCasedObj[camelCasedKey] = obj[key];
  });

  return camelCasedObj;
};
