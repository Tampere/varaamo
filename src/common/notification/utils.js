import { NotificationManager } from 'react-notifications';

import { NOTIFICATION_TYPE } from './constants';
/**
 * createNotification()
 * Creating notification popups
 *
 * @param {string} type
 * @param {string} [message='']
 * @param {string} [title='']
 * @param {number} [fadeTime=2000]
 * @param {function} callback
 * @returns {void}
 */
export const createNotification = (type, message = '', title = '', fadeTime = 2000, callback) => {
  switch (type) {
    case NOTIFICATION_TYPE.INFO:
      NotificationManager.info(message, title, fadeTime, callback);
      break;
    case NOTIFICATION_TYPE.SUCCESS:
      NotificationManager.success(message, title, fadeTime, callback);
      break;
    case NOTIFICATION_TYPE.WARNING:
      NotificationManager.warning(message, title, fadeTime, callback);
      break;
    case NOTIFICATION_TYPE.ERROR:
      NotificationManager.error(message, title, fadeTime, callback);
      break;
    default:
      NotificationManager.info(message, title, fadeTime, callback);
  }
};
