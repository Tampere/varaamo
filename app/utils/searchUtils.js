import constants from 'constants/AppConstants';

import omit from 'lodash/omit';
import pick from 'lodash/pick';
import queryString from 'query-string';

import { getDateStartAndEndTimes, getDateString } from 'utils/timeUtils';

function getFetchParamsFromFilters(filters) {
  const all = Object.assign(
    {},
    pickSupportedFilters(filters),
    getDateStartAndEndTimes(
      filters.date,
      filters.useTimeRange,
      filters.start,
      filters.end,
      filters.duration
    ),
    { purpose: filters.purpose === 'all' ? '' : filters.purpose },
    { page: filters.page || 1 }
  );

  return omit(all, 'date', 'duration', 'useTimeRange');
}

function getSearchPageUrl(filters = {}) {
  const query = queryString.stringify(
    Object.assign({}, filters, { date: getDateString(filters.date) })
  );

  return `/search?${query}`;
}

function pickSupportedFilters(filters) {
  return pick(filters, Object.keys(constants.SUPPORTED_SEARCH_FILTERS));
}

function textBoolean(value) {
  return value === 'true' || value === true;
}

export {
  getFetchParamsFromFilters, getSearchPageUrl, pickSupportedFilters, textBoolean
};
