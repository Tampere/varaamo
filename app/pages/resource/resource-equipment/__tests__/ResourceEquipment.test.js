import React from 'react';
import toJSON from 'enzyme-to-json';

import { shallowWithIntl } from '../../../../utils/testUtils';
import ResourceEquipment from '../ResourceEquipment';

test('pages/resource/resource-equipment/ResourceEquipment', () => {
  const props = {
    equipment: [
      {
        name: 'Karaoke',
      },
      {
        name: 'Projector',
      },
      {
        name: 'Printer',
      },
    ],
  };


  const wrapper = shallowWithIntl(
    <ResourceEquipment {...props} />,
  );

  expect(toJSON(wrapper)).toMatchSnapshot();
});
