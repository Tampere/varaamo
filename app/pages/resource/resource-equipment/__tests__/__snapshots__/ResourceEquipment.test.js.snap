// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/resource/resource-equipment/ResourceEquipment 1`] = `
<ResourcePanel
  header="ResourceEquipment.headingText"
>
  <Row
    bsClass="row"
    componentClass="div"
  >
    <Col
      bsClass="col"
      componentClass="div"
      key="0"
      lg={3}
      md={3}
      xs={6}
    >
      Karaoke
    </Col>
    <Col
      bsClass="col"
      componentClass="div"
      key="1"
      lg={3}
      md={3}
      xs={6}
    >
      Projector
    </Col>
    <Col
      bsClass="col"
      componentClass="div"
      key="2"
      lg={3}
      md={3}
      xs={6}
    >
      Printer
    </Col>
  </Row>
</ResourcePanel>
`;
